l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol

```diff
@@ -0,0 +1,201 @@
+    /// @param _l2TokenProxyBytecodeHash The bytecode hash of the proxy for tokens deployed by the bridge.
+    /// @param _aliasedOwner The address of the governor contract.
+    /// @param _legacySharedBridge The address of the L2 legacy shared bridge.
+    /// @param _bridgedTokenBeacon The address of the L2 token beacon for legacy chains.
+    /// @param _contractsDeployedAlready Ensures beacon proxy for standard ERC20 has not been deployed.
+    /// @param _wethToken Address of WETH on deployed chain
+    constructor(
+        uint256 _l1ChainId,
+        address _aliasedOwner,
+        bytes32 _l2TokenProxyBytecodeHash,
+        address _legacySharedBridge,
+        address _bridgedTokenBeacon,
+        bool _contractsDeployedAlready,
+        address _wethToken
+    ) NativeTokenVault(_wethToken, L2_ASSET_ROUTER_ADDR) {
+        L1_CHAIN_ID = _l1ChainId;
+        L2_LEGACY_SHARED_BRIDGE = IL2SharedBridgeLegacy(_legacySharedBridge);
+
+        _disableInitializers();
+        if (_l2TokenProxyBytecodeHash == bytes32(0)) {
+            revert EmptyBytes32();
+        }
+        if (_aliasedOwner == address(0)) {
+            revert EmptyAddress();
+        }
+
+        l2TokenProxyBytecodeHash = _l2TokenProxyBytecodeHash;
+        _transferOwnership(_aliasedOwner);
+
+        if (_contractsDeployedAlready) {
+            if (_bridgedTokenBeacon == address(0)) {
+                revert EmptyAddress();
+            }
+            bridgedTokenBeacon = IBeacon(_bridgedTokenBeacon);
+        } else {
+            address l2StandardToken = address(new BridgedStandardERC20{salt: bytes32(0)}());
+
+            UpgradeableBeacon tokenBeacon = new UpgradeableBeacon{salt: bytes32(0)}(l2StandardToken);
+
+            tokenBeacon.transferOwnership(owner());
+            bridgedTokenBeacon = IBeacon(address(tokenBeacon));
+            emit L2TokenBeaconUpdated(address(bridgedTokenBeacon), _l2TokenProxyBytecodeHash);
+        }
+    }
+
+    /// @notice Sets the legacy token asset ID for the given L2 token address.
+    function setLegacyTokenAssetId(address _l2TokenAddress) public {
+        address l1TokenAddress = L2_LEGACY_SHARED_BRIDGE.l1TokenAddress(_l2TokenAddress);
+        bytes32 assetId = DataEncoding.encodeNTVAssetId(L1_CHAIN_ID, l1TokenAddress);
+        tokenAddress[assetId] = _l2TokenAddress;
+    }
+
+    /// @notice Ensures that the token is deployed.
+    /// @param _originChainId The chain ID of the origin chain.
+    /// @param _assetId The asset ID.
+    /// @param _originToken The origin token address.
+    /// @param _erc20Data The ERC20 data.
+    /// @return _token The token address.
+    function _ensureTokenDeployed(
+        uint256 _originChainId,
+        bytes32 _assetId,
+        address _originToken,
+        bytes memory _erc20Data
+    ) internal override returns (address _token) {
+        address expectedToken = _assetIdCheck(_originChainId, _assetId, _originToken);
```

Comment:
nit: I think we could just asset `_token` here rightaway. Also, by convention the returned values should not start with underscore